lgc/patch/PatchReadfirstlaneOpt.cpp

```diff
@@ -0,0 +1,232 @@
+
+  static char ID; // ID of this pass
+
+private:
+  PatchReadfirstlaneOpt(const PatchReadfirstlaneOpt &) = delete;
+  PatchReadfirstlaneOpt &operator=(const PatchReadfirstlaneOpt &) = delete;
+
+  bool pushUpReadfirstlane();
+
+  LegacyDivergenceAnalysis *m_divergenceAnalysis; // The divergence analysis.
+  DenseSet<Value *> m_initialUniforms;            // The initial uniforms set for starting readfirstlane optimization
+};
+
+} // namespace lgc
+
+// =====================================================================================================================
+// Initializes static members.
+char lgc::PatchReadfirstlaneOpt::ID = 0;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of LLVM patching operations for readfirstlane optimizations.
+FunctionPass *lgc::createPatchReadfirstlaneOpt() {
+  return new PatchReadfirstlaneOpt();
+}
+
+PatchReadfirstlaneOpt::PatchReadfirstlaneOpt() : FunctionPass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in,out] function : Function that we will optimize readfirstlane insertion place.
+bool PatchReadfirstlaneOpt::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Readfirstlane-Opt\n");
+
+  m_divergenceAnalysis = &getAnalysis<LegacyDivergenceAnalysis>();
+  visit(function);
+
+  const bool changed = pushUpReadfirstlane();
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Specify what analysis passes this pass depends on.
+//
+// @param [in,out] analysisUsage : The place to record our analysis pass usage requirements.
+void PatchReadfirstlaneOpt::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
+  analysisUsage.addRequired<LegacyDivergenceAnalysis>();
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
+//
+// @param callInst : The instruction
+void PatchReadfirstlaneOpt::visitCallInst(CallInst &callInst) {
+  // Collect the intial uniform operands as the start point for pushing the corresponding readfirstlanes higher
+  if (auto intrinsicCall = dyn_cast<IntrinsicInst>(&callInst)) {
+    if (intrinsicCall->getIntrinsicID() == Intrinsic::amdgcn_readfirstlane) {
+      // If readfirstlane is placed on each element descriptor, use extract to only add the descriptor to the set
+      Value *const readfirstlaneArg = callInst.getArgOperand(0);
+      if (auto extract = dyn_cast<ExtractElementInst>(readfirstlaneArg))
+        m_initialUniforms.insert(extract->getOperand(0));
+      else if (!dyn_cast<CallInst>(readfirstlaneArg))
+        m_initialUniforms.insert(readfirstlaneArg);
+    }
```

Comment:
Remove the limit. Checking for extractElement is to avoid adding reduncant uniforms for descriptor case. If the readfirstlane argument is a callinst, there is no need to insert a readfirstlane on it.